compiler/src/dotty/tools/dotc/transform/Pickler.scala

```diff
@@ -33,8 +38,62 @@ object Pickler {
    */
   inline val ParallelPickling = true
 
+  class AsyncTastyHolder(val earlyOut: AbstractFile, val promise: Promise[AsyncTastyState])
+  class AsyncTastyState(val hasErrors: Boolean, val pending: Option[BufferingDelayedReporting])
+
+  // Why we only write to early output in the first run?
+  // ===================================================
+  // TL;DR the point of pipeline compilation is to start downstream projects early,
+  // so we don't want to wait for suspended units to be compiled.
+  //
+  // But why is it safe to ignore suspended units?
+  // If this project contains a transparent macro that is called in the same project,
+  // the compilation unit of that call will be suspended (if the macro implementation
+  // is also in this project), causing a second run.
+  // However before we do that run, we will have already requested sbt to begin
+  // early downstream compilation. This means that the suspended definitions will not
+  // be visible in *early* downstream compilation.
+  //
+  // However, sbt will by default prevent downstream compilation happening in this scenario,
+  // due to the existence of macro definitions. So we are protected from failure if user tries
+  // to use the suspended definitions.
+  //
+  // Additionally, it is recommended for the user to move macro implementations to another project
+  // if they want to force early output. In this scenario the suspensions will no longer occur, so now
+  // they will become visible in the early-output.
+  //
+  // See `sbt-test/pipelining/pipelining-scala-macro` and `sbt-test/pipelining/pipelining-scala-macro-force`
+  // for examples of this in action.
+  //
+  // Therefore we only need to write to early output in the first run. We also provide the option
+  // to diagnose suspensions with the `-Yno-suspended-units` flag.
+  def writeSigFilesAsync(
+      tasks: List[(String, Array[Byte])],
+      writer: EarlyFileWriter,
+      promise: Promise[AsyncTastyState])(using ctx: ReadOnlyContext): Unit = {
+    try
+      for (internalName, pickled) <- tasks do
+        val _ = writer.writeTasty(internalName, pickled)
+    finally
+      try
+        writer.close()
+      finally
```

Comment:
That's pretty weird. It's hard to reason about exception handling nested inside a `finally` block. More so if it is itself a `try..finally`.

Consider structuring as follows instead:
```scala
try
  try
    for ... do
      writer.writeTasty(...)
  finally
    writer.close()
  end try
finally
  promise.succeed(...)
end try
```
Even then, it seems weird to complete a promise with `success` inside a `finally` block. Is there any reason that we need to complete the promise at all if an exception gets raised here? And if yes, why not completing it with an actual `Failure` that contains the thrown exception?

Reply:
yeah its a good point. The reason to always succeed is so that the buffered logs will eventually be sent to the backend. So I guess any exceptions should log to the (buffered) reporter, but then still always succeed

Reply:
implemented the re-ordering of try, and now capture any non-fatal throwable in the logs